File: Max Number of K Sum Pairs.py
from collections import Counter
from typing import List

class Solution:
    def maxOperations(self, nums: List[int], k: int) -> int:
        counts = Counter(nums)
        operations = 0
        seen = set()
        for key in counts:
            if key in seen:
                continue

            if k - key == key:
                operations += counts[key] // 2
                seen.add(key)
            elif k - key in counts:
                operations += min(counts[key], counts[k-key])
                seen.add(key)
                seen.add(k-key)
        return operations

File: cpp/Max Number of K Sum Pairs.cpp
#include <vector>
#include <unordered_map>
#include <iostream>
#include <algorithm>
using namespace std;


class Solution {
public:
    int maxOperations(vector<int>& nums, int k) {
        unordered_map<int, int> counter;
        int k_pairs = 0;

        for (const auto n: nums) {
            if (counter.count(n) > 0) {
                counter[n]++;
            } else {
                counter[n] = 1;
            }
        }

        for (const auto& pair: counter) {
            if (pair.first * 2 == k) {
                k_pairs += (pair.second / 2) * 2;
            } else if (counter.count(k-pair.first) > 0) {
                k_pairs += min(pair.second, counter[k-pair.first]);
            }
        }
        return k_pairs/2;
    }
};